aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -186,6 +186,70 @@ Tensor ger(const Tensor& self, const Tensor& vec2) {
   return self.outer(vec2);
 }
 
+Tensor& inner_out(Tensor& out, const Tensor& self, const Tensor& other) {
+  checkDeviceType("inner()", {out, self, other}, self.device().type());
+
+  // If either self or other is a scalar just multiply them
+  if (self.dim() == 0 || other.dim() == 0) {
+    resize_output(out, self.dim() == 0 ? other.sizes() : self.sizes());
+    at::mul_out(out, self, other);
+    return out;
+  }
+
+  // Last dimension should match
+  TORCH_CHECK(
+      self.size(-1) == other.size(-1),
+      "inner() the last dimension must match on both input tensors but got shapes ",
+      self.sizes(),
+      " and ",
+      other.sizes());
+
+  // Out shape is self dims (except last) + other dims (except last)
+  std::vector<int64_t> out_shape(
+      self.sizes().begin(), self.sizes().begin() + self.dim() - 1);
+  out_shape.insert(
+      out_shape.end(),
+      other.sizes().begin(),
+      other.sizes().begin() + other.dim() - 1);
+  resize_output(out, out_shape);
+
+  // Broadcast all but the last dimension
+  Tensor input = self;
+  for (int64_t i = other.dim() - 1; i > 0; --i) {
+    input = input.unsqueeze(-2);
+  }
+
+  // Compute result
+  at::sum_out(out, input * other, -1);
+  return out;
+}
+
+Tensor inner(const Tensor& self, const Tensor& other) {
+  checkDeviceType("inner()", {self, other}, self.device().type());
+
+  // If either self or other is a scalar just multiply them
+  if (self.dim() == 0 || other.dim() == 0) {
+    return self * other;
+  }
+
+  // Last dimension should match
+  TORCH_CHECK(
+      self.size(-1) == other.size(-1),
+      "inner() the last dimension must match on both input tensors but got shapes ",
+      self.sizes(),
+      " and ",
+      other.sizes());
+
+  // Broadcast all but the last dimension
+  Tensor input = self;
+  for (int64_t i = other.dim() - 1; i > 0; --i) {
+    input = input.unsqueeze(-2);
```

Comment:
Same question here as in the `out` implementation. How does this unsqueeze happening multiple times relate to broadcasting?